### PR TITLE
Refactor navigation destinations

### PR DESCRIPTION
**Pull request by mplorentz**
## Issues covered
This is the beginning of the solution to problem 2 in https://github.com/planetary-social/nos/issues/703#issuecomment-2147829931
> Using an NSManagedObject as a SwiftUI .navigationDestination breaks object observation. Solution: use IDs as the navigation destinations and set up container views that observe the object with the give ID.

## Description
This is purely a refactoring PR. The app should work exactly the same after this is merged. Each tab in the app and the side menu all had duplicated code for handling `navigationDestination()`s. This refactors them into a shared view. In the next PR I'll change how the destinations work.

## How to test
Verify that each tab can still push a detail view for each type of content - Event, Author, URL, etc. Pay special attention to the Profile tab and side menu - these had some duplicated navigation destinations that I removed.

### CODE REVIEW

File: Nos/Views/Components/NostrNavigationStack.swift [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import SwiftUI
+
+/// A NavigationStack that knows how to present views to display Nostr entities like `Events` and `Authors`.
+/// Take care not to nest these.
+struct NostrNavigationStack<Content: View>: View {
```
Comment (martindsq) [inline Nos/Views/Components/NostrNavigationStack.swift:5]: I'm not very convinced about this name. It doesn't look precise to me. The fact that it also adds URL and ReplyToNavigationDestination makes me think that it is more Nos related than Nostr.

File: Nos/Views/Discover/DiscoverTab.swift [hunk 2]
```diff
@@ -65,19 +65,6 @@ struct DiscoverTab: View {
...
-                RepliesView(note: note)
-            }
-            .navigationDestination(for: URL.self) { url in URLView(url: url) }
-            .navigationDestination(for: ReplyToNavigationDestination.self) { destination in 
-                RepliesView(note: destination.note, showKeyboard: true)
-            }
-            .navigationDestination(for: Author.self) { author in
-                ProfileView(author: author)
-            }
-            .navigationDestination(for: EditProfileDestination.self) { destination in
```
Comment (martindsq) [inline Nos/Views/Discover/DiscoverTab.swift:78]: We are losing this one, right? Maybe its fine, I don't see how EditProfile can be directly opened from the DiscoverTab.

File: Nos/Views/Home/HomeTab.swift [hunk 3]
```diff
@@ -6,24 +6,10 @@ struct HomeTab: View {
...
-        NavigationStack(path: $router.homeFeedPath) {
+        NostrNavigationStack(path: $router.homeFeedPath) {
             HomeFeedView(user: user)
-                .navigationDestination(for: Event.self) { note in
-                    RepliesView(note: note)
-                }
-                .navigationDestination(for: Author.self) { author in
-                    ProfileView(author: author)
-                }
-                .navigationDestination(for: EditProfileDestination.self) { destination in
```
Comment (martindsq) [inline Nos/Views/Home/HomeTab.swift:20]: Same for this one, I wondered why we added these in first place.

File: Nos/Views/NotificationsView.swift [hunk 4]
```diff
@@ -90,19 +90,6 @@ struct NotificationsView: View {
...
-                RepliesView(note: note)
-            }
-            .navigationDestination(for: URL.self) { url in URLView(url: url) }
-            .navigationDestination(for: ReplyToNavigationDestination.self) { destination in
-                RepliesView(note: destination.note, showKeyboard: true)
-            }
-            .navigationDestination(for: Author.self) { author in
-                ProfileView(author: author)
-            }
-            .navigationDestination(for: EditProfileDestination.self) { destination in
```
Comment (martindsq) [inline Nos/Views/NotificationsView.swift:103]: And this one, maybe it was just in case?

**Review by martindsq — COMMENTED** [review]
We are losing `.navigationDestination(for: EditProfileDestination.self)` in all cases, but it seems safe to remove. We shouldn't be opening the EditProfile directly from these cases. I tested it to be double-sure and opening the Edit Profile from the banner that appears when you create an account still works.

I just left a comment about NostrNavigationStack. I think it should be NosNavigationStack because it includes more than NostrEntities. We can ask Josh to give advice about this.

File: Nos/Views/Components/NostrNavigationStack.swift (hunk 1 above)
Comment (joshuatbrown) [reply Nos/Views/Components/NostrNavigationStack.swift:5]: I agree with your other comment; I like NosNavigationStack for the name here. But I'm open to NostrNavigationStack if @mplorentz has a convincing reason.

Comment (mplorentz) [reply Nos/Views/Components/NostrNavigationStack.swift]: I was already on the fence between `NosNavigationStack` and `NostrNavigationStack`. Updated!

Comment (mplorentz) [reply Nos/Views/Components/NostrNavigationStack.swift]: Ah, I just remembered what pushed me towards `NostrNavigationStack` was that I'm creating a `NostrNavigationDestination` enum in my next PR, and `NosNavigationDestination` sounds like it's every navigation destination in the app, which it isn't. But we can address that in the next PR.

**Review by martindsq — APPROVED** [review]

**Pull request merged**